test/e2e/scale.go

```diff
@@ -0,0 +1,577 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/onsi/ginkgo/v2/types"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/e2e/internal/log"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util/yaml"
+)
+
+const (
+	scaleClusterCount             = "CAPI_SCALE_CLUSTER_COUNT"
+	scaleConcurrency              = "CAPI_SCALE_CONCURRENCY"
+	scaleControlPlaneMachineCount = "CAPI_SCALE_CONTROL_PLANE_MACHINE_COUNT"
+	scaleWorkerMachineCount       = "CAPI_SCALE_WORKER_MACHINE_COUNT"
+)
+
+// scaleSpecInput is the input for scaleSpec.
+type scaleSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+
+	// InfrastructureProvider allows to specify the infrastructure provider to be used when looking for
+	// cluster templates.
+	// If not set, clusterctl will look at the infrastructure provider installed in the management cluster;
+	// if only one infrastructure provider exists, it will be used, otherwise the operation will fail if more than one exists.
+	InfrastructureProvider *string
+
+	// Flavor, if specified is the template flavor used to create the cluster for testing.
+	// If not specified, the default flavor for the selected infrastructure provider is used.
+	Flavor *string
+
+	// ClusterCount is the number of target workload clusters.
+	// Can be overridden by env variable CAPI_SCALE_CLUSTER_COUNT.
+	ClusterCount int
+
+	// Concurrency is the maximum concurrency of each of the scale operations.
+	// If unspecified it defaults to 1.
+	// Can be overridden by env variable CAPI_SCALE_CONCURRENCY.
+	Concurrency int
+
+	// ControlPlaneMachineCount defines the number of control plane machines to be added to each workload cluster.
+	// If not specified, 1 will be used.
+	// Can be overridden by env variable CAPI_SCALE_CONTROLPLANE_MACHINE_COUNT.
+	ControlPlaneMachineCount *int64
+
+	// WorkerMachineCount defines number of worker machines to be added to each workload cluster.
+	// If not specified, 1 will be used.
+	// Can be overridden by env variable CAPI_SCALE_WORKER_MACHINE_COUNT.
+	WorkerMachineCount *int64
+
+	// Allows to inject functions to be run while waiting for the control plane to be initialized,
+	// which unblocks CNI installation, and for the control plane machines to be ready (after CNI installation).
+	ControlPlaneWaiters clusterctl.ControlPlaneWaiters
```

Comment:
Wondering if we should drop this for now

Reply:
Just seems a bit arbtirary to add this right now wiithout a clear use case (vs all the other "hooks" that we have in other tests)